client/my-sites/pages/page-list.jsx

```diff
@@ -203,37 +203,40 @@ var Pages = React.createClass( {
 
 		// pages have loaded, sites have loaded, and we have a site instance or are viewing all-sites
 		if ( pages.length && this.props.sites.initialized ) {
-			if ( ! this.props.search ) {
-				// we're listing in reverse chrono. use the markers.
-				pages = this._insertTimeMarkers( pages );
-			}
-			rows = pages.map( function( page ) {
-				if ( ! ( 'site_ID' in page ) ) {
-					return page;
+			const site = this.props.sites.getSelectedSite();
+			const status = this.props.status || 'published';
+
+			// @TODO check if it is the first and last page (means there are fewer than 100)
+			if ( site && status === 'published' && ! this.props.search && this.props.lastPage ) {
+				pages = sortPagesHierarchically( pages );
+
+				rows = pages.map( function( page ) {
+					return (
+						<Page key={ 'page-' + page.global_ID } page={ page } site={ site }
+							multisite={ false } hierarchical={ true } hierarchyLevel={ page.indentLevel || 0 } />
+					);
+				}, this );
+			} else {
+				if ( ! this.props.search ) {
+					// we're listing in reverse chrono. use the markers.
+					pages = this._insertTimeMarkers( pages );
 				}
+				rows = pages.map( function( page ) {
+					if ( ! ( 'site_ID' in page ) ) {
+						return page;
+					}
 					// Get the site the page belongs to
-				var site = this.props.sites.getSite( page.site_ID );
+					const pageSite = this.props.sites.getSite( page.site_ID );
 
 					// Render each page
-				return (
-						<Page key={ 'page-' + page.global_ID } page={ page } site={ site } multisite={ this.props.siteID === false } />
+					return (
+						<Page key={ 'page-' + page.global_ID } page={ page } site={ pageSite } multisite={ this.props.siteID === false } />
 					);
-			}, this );
-
-			if ( this.props.loading ) {
-				this.addLoadingRows( rows, 1 );
-			}
+				}, this );
 
-			const site = this.props.sites.getSelectedSite();
-			const status = this.props.status || 'published';
-
-			if ( site && status === 'published' ) {
-				rows.push(
-					<BlogPostsPage
-						key="blog-posts-page"
-						site={ site }
-					/>
-				);
+				if ( this.props.loading ) {
+					this.addLoadingRows( rows, 1 );
+				}
```

Comment:
I'd really work on getting rid of the current rendering model here, which looks like:

```js
var rows;

if ( A ) {
  …
  rows.push( … );
} else {
  if ( B & C ) {
    rows.push( … );
  } else if ( D ) {
    rows.push( … );
    this.addLoadingRows( rows, n ); // mutates `rows`!
  }
} else if ( E ) {
  … etc
}

return (
  <scaffolding>
    { rows }
  </scaffolding>
);
```

and would either have it like

```js
if ( C ) {
  return <SomePlaceholderForPages />;
}

if ( B ) {
  return <SomeErrorHandlingOfPages />;
}

if ( A ) {
  return <SomeVersionOfPages>SomeData</SomeVersionOfPages>;
}
```

so that:
- there's less state to keep track of along the method
- there's a direct relationship between specific conditions and return value; could be achieved with a `switch` block too
- there's a better approximation of the classic model of 3 possible render outcomes: "data", "no data", "waiting for data"